File: android-native/app/src/main/java/com/poverse/app/data/model/Attendance.kt
package com.poverse.app.data.model

import kotlinx.serialization.Serializable

@Serializable
data class AttendanceRecord(
    val id: String = "",
    val userId: String = "",
    val companyId: String = "",
    val date: String = "", // YYYY-MM-DD
    val checkIn: CheckInData? = null,
    val checkOut: CheckOutData? = null,
    val status: AttendanceStatus = AttendanceStatus.ABSENT,
    val lateStatus: LateStatus = LateStatus.ON_TIME,
    val earlyDepartureStatus: String = "",
    val workDurationMinutes: Int = 0,
    val createdAt: Long = System.currentTimeMillis()
)

@Serializable
data class CheckInData(
    val timestamp: Long = 0,
    val location: LocationPoint? = null,
    val selfieUrl: String = "",
    val address: String = "",
    val deviceInfo: String = ""
)

@Serializable
data class CheckOutData(
    val timestamp: Long = 0,
    val location: LocationPoint? = null,
    val selfieUrl: String = "",
    val address: String = "",
    val deviceInfo: String = "",
    val autoCheckout: Boolean = false
)

@Serializable
data class LocationPoint(
    val latitude: Double = 0.0,
    val longitude: Double = 0.0,
    val accuracy: Float = 0f,
    val address: String = ""
)

@Serializable
enum class AttendanceStatus {
    CHECKED_IN, CHECKED_OUT, ABSENT, ON_LEAVE, HALF_DAY;

    companion object {
        fun fromString(value: String): AttendanceStatus = when (value.lowercase()) {
            "checked_in" -> CHECKED_IN
            "checked_out" -> CHECKED_OUT
            "on_leave" -> ON_LEAVE
            "half_day" -> HALF_DAY
            else -> ABSENT
        }
    }

    override fun toString(): String = when (this) {
        CHECKED_IN -> "checked_in"
        CHECKED_OUT -> "checked_out"
        ABSENT -> "absent"
        ON_LEAVE -> "on_leave"
        HALF_DAY -> "half_day"
    }
}

@Serializable
enum class LateStatus {
    ON_TIME, LATE, VERY_LATE;

    companion object {
        fun fromString(value: String): LateStatus = when (value.lowercase()) {
            "late" -> LATE
            "very_late" -> VERY_LATE
            else -> ON_TIME
        }
    }

    override fun toString(): String = when (this) {
        ON_TIME -> "on_time"
        LATE -> "late"
        VERY_LATE -> "very_late"
    }
}

@Serializable
data class AttendanceSettings(
    val companyId: String = "",
    val shiftStartTime: String = "09:00",
    val shiftEndTime: String = "18:00",
    val lateThresholdMinutes: Int = 15,
    val veryLateThresholdMinutes: Int = 30,
    val autoCheckoutEnabled: Boolean = true,
    val autoCheckoutTime: String = "20:00",
    val selfieRequired: Boolean = true,
    val locationRequired: Boolean = true,
    val workingDays: List<Int> = listOf(1, 2, 3, 4, 5) // Mon-Fri
)

File: src/types/attendance.ts
// Attendance system types

export type AttendanceStatus = 
  | "checked_in"      // User has checked in
  | "checked_out"     // User has checked out
  | "absent"          // User didn't check in
  | "on_leave"        // User is on approved leave
  | "half_day";       // User worked half day

export type LateStatus = 
  | "on_time"         // Checked in before shift start
  | "late"            // Checked in after shift start
  | "very_late";      // Checked in significantly late (>30 mins)

export type EarlyDepartureStatus = 
  | "full_day"        // Checked out after shift end
  | "early"           // Checked out before shift end
  | "very_early";     // Checked out significantly early (>1 hour)

export interface AttendanceLocation {
  latitude: number;
  longitude: number;
  accuracy: number | null;
  address?: string;
}

export interface CheckInData {
  location: AttendanceLocation;
  selfieUrl: string;
  selfieStoragePath: string;
  timestamp: string;
  deviceInfo?: string;
}

export interface CheckOutData {
  location: AttendanceLocation;
  timestamp: string;
  deviceInfo?: string;
  notes?: string;
}

export interface AttendanceRecord {
  id: string;
  userId: string;
  userName: string;
  companyId: string;
  date: string; // YYYY-MM-DD format
  
  // Check-in details
  checkIn?: CheckInData;
  
  // Check-out details
  checkOut?: CheckOutData;
  
  // Status tracking
  status: AttendanceStatus;
  lateStatus?: LateStatus;
  earlyDepartureStatus?: EarlyDepartureStatus;
  
  // Time calculations
  workDurationMinutes?: number;
  lateByMinutes?: number;
  earlyByMinutes?: number;
  
  // Metadata
  createdAt: string;
  updatedAt: string;
  
  // Auto checkout flag
  isAutoCheckedOut?: boolean;
  autoCheckoutReason?: string;
}

export interface AttendanceSettings {
  companyId: string;
  shiftStartTime: string;  // HH:mm format (e.g., "09:00")
  shiftEndTime: string;    // HH:mm format (e.g., "18:00")
  lateThresholdMinutes: number;     // Minutes after shift start to be marked late
  veryLateThresholdMinutes: number; // Minutes after shift start to be marked very late
  earlyDepartureThresholdMinutes: number;  // Minutes before shift end for early departure
  autoCheckoutTime: string; // HH:mm format (e.g., "23:59")
  autoCheckoutEnabled: boolean;
  selfieRequired: boolean;
  locationRequired: boolean;
  workingDays: number[];   // 0 = Sunday, 1 = Monday, etc.
}

export interface DailyAttendanceSummary {
  date: string;
  totalEmployees: number;
  present: number;
  absent: number;
  onLeave: number;
  late: number;
  earlyDeparture: number;
  averageWorkHours: number;
}

export interface MonthlyAttendanceSummary {
  userId: string;
  userName: string;
  month: string; // YYYY-MM format
  totalWorkingDays: number;
  presentDays: number;
  absentDays: number;
  leaveDays: number;
  lateDays: number;
  earlyDepartureDays: number;
  totalWorkHours: number;
  averageWorkHours: number;
  attendancePercentage: number;
}

export interface AttendanceCalendarDay {
  date: string;
  status: AttendanceStatus | "weekend" | "holiday" | "future";
  checkInTime?: string;
  checkOutTime?: string;
  workHours?: number;
  isLate?: boolean;
  isEarlyDeparture?: boolean;
}

// Default attendance settings
export const DEFAULT_ATTENDANCE_SETTINGS: Omit<AttendanceSettings, "companyId"> = {
  shiftStartTime: "09:00",
  shiftEndTime: "18:00",
  lateThresholdMinutes: 15,
  veryLateThresholdMinutes: 30,
  earlyDepartureThresholdMinutes: 60,
  autoCheckoutTime: "23:59",
  autoCheckoutEnabled: true,
  selfieRequired: true,
  locationRequired: true,
  workingDays: [1, 2, 3, 4, 5, 6], // Monday to Saturday
};
